pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ForkliftPopulatorReconciler) updateAnnotations(pvc, pvcPrime *corev1.PersistentVolumeClaim) {
+	phase := pvcPrime.Annotations[cc.AnnPodPhase]
+	if err := r.updateImportProgress(phase, pvc, pvcPrime); err != nil {
+		r.log.Error(err, "Failed to update import progress for pvc %s/%s", pvc.Namespace, pvc.Name)
+	}
+}
+
+func (r *ForkliftPopulatorReconciler) updatePVCForPopulation(pvc *corev1.PersistentVolumeClaim, source client.Object) {
+	annotations := pvc.Annotations
+	annotations[cc.AnnUsePopulator] = "true"
+	cc.AddAnnotation(pvc, cc.AnnPopulatorKind, "forklift")
+	cc.AddAnnotation(pvc, cc.AnnUsePopulator, "true")
+}
+
+func (r *ForkliftPopulatorReconciler) updateImportProgress(podPhase string, pvc, pvcPrime *corev1.PersistentVolumeClaim) error {
+	// Just set 100.0% if pod is succeeded
+	if podPhase == string(corev1.PodSucceeded) {
+		cc.AddAnnotation(pvc, cc.AnnPopulatorProgress, "100.0%")
+		return nil
+	}
+
+	importPodName, ok := pvcPrime.Annotations[cc.AnnImportPod]
+	if !ok {
+		return nil
+	}
+
+	importPod, err := r.getImportPod(pvcPrime, importPodName)
+	if err != nil {
+		return err
+	}
+
+	if importPod == nil {
+		_, ok := pvc.Annotations[cc.AnnPopulatorProgress]
+		// Initialize the progress once PVC Prime is bound
+		if !ok && pvcPrime.Status.Phase == corev1.ClaimBound {
+			cc.AddAnnotation(pvc, cc.AnnPopulatorProgress, "N/A")
+		}
+		return nil
+	}
+
+	// This will only work when the import pod is running
+	if importPod.Status.Phase != corev1.PodRunning {
+		return nil
+	}
+
+	url, err := cc.GetMetricsURL(importPod)
+	if url == "" || err != nil {
+		return err
+	}
+
+	// We fetch the import progress from the import pod metrics
+	importRegExp := regexp.MustCompile("progress\\{ownerUID\\=\"" + string(pvc.UID) + "\"\\} (\\d{1,3}\\.?\\d*)")
+	httpClient = cc.BuildHTTPClient(httpClient)
+	progressReport, err := cc.GetProgressReportFromURL(url, importRegExp, httpClient)
+	if err != nil {
+		return err
+	}
+
+	if progressReport != "" {
+		if f, err := strconv.ParseFloat(progressReport, 64); err == nil {
+			cc.AddAnnotation(pvc, cc.AnnPopulatorProgress, fmt.Sprintf("%.2f%%", f))
+		}
+	}
+
+	return nil
+}
+
+func (r *ForkliftPopulatorReconciler) getImportPod(pvc *corev1.PersistentVolumeClaim, importPodName string) (*corev1.Pod, error) {
+	pod := &corev1.Pod{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: importPodName, Namespace: pvc.GetNamespace()}, pod); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, nil
+	}
+
+	if !metav1.IsControlledBy(pod, pvc) && !cc.IsImageStream(pvc) {
+		return nil, errors.Errorf("Pod is not owned by PVC")
+	}
+	return pod, nil
+}
+
+func (r *ForkliftPopulatorReconciler) createPopulatorPod(pvcPrime, pvc *corev1.PersistentVolumeClaim) error {
+	var rawBlock bool
+	if nil != pvc.Spec.VolumeMode && corev1.PersistentVolumeBlock == *pvc.Spec.VolumeMode {
```

Comment:
I am surprised the linter is not unhappy with this. Can we flip the arguments to `pvc.Spec.VolumeMode != nil && *pvc.Spec.VolumeMode == corev1.PersistentVolumeBlock`?